### DISCUSSION TITLE
Disabling Pylance Linting, while keeping Pyright Type Checking.

### DISCUSSION THREAD

**Discussion by farzbood** [microsoft/pylance-release#6492] [Q&A]
Hi,
I'm using **Ruff** as my **Python Linter**, so don't need **Pylance Linting** functionality, but want to keep using **PyRight** as static **Type-Checking** tool.

I've tried many settings/configurations to achieve that but no success yet!
Why are these Language Server/Linter and Type Checker so tight coupled anyway?!
Is there a way to do so or should just use another tool e.g: mypy instead?

Any suggestions would be appreciated.

**Comment by rchiodo** [comment]
Linting is kind of subjective, but if you set 'typeCheckingMode' to 'off', that should eliminate all of the errors that Pylance shows besides syntax errors. But intellisense and everything else would still work.

**Reply by rchiodo** [reply]
The `python.analysis.ignore` setting can also achieve a similar result. It ignores errors for everything that matches the ignore glob.

**Comment by farzbood** [comment]
I'm not sure have you got my point or even try your answer either, or what?
### Bypassing Linting (error, warnings, etc.) while keeping Type-Checking (Type annotation/hint) in place and the suggestion is setting 'typeCheckingMode' to 'off' ?! really? (it's the other way and the default)
`python.analysis.ignore` is about **Paths** to analyze, not the way of treating the results, setting it to '*' (which is the only way to bypass Linting) supress everything (including Type-Checking)!

My guess is you at Pylance, think of **Code Analysis** in **Python**(a dynamic typed scripting language) as of a strong typed language that is not right.
What is Pylance anyway? Pylint + PyRight + ... that each of which belong to different use-cases and need to be manageable separately. remember (SoC).

**Reply by rchiodo** [reply]
Pylint is not involved in Pylance. Pylance solely uses Pyright to generate and check types. There's a bunch of settings that let you fine tune what errors are generated. They're documented here:

https://microsoft.github.io/pyright/#/configuration?id=type-check-diagnostics-settings

For pylance there's no such thing as 'linting' rules vs 'typechecking' rules, so I thought you wanted to turn them all off and just have Ruff generate errors. 

Turning typeCheckingMode to off has no impact on hover/completions/inlay hints, they work the same no matter what this setting is.

**Reply by farzbood** [reply]
IMO, there are three different domains in the realm of **Code analysis**.
**1) Language specific rules** (syntax errors, warnings, etc.) that are mandatory for code to function correctly (otherwise breaks at runtime)
**2) Type-hint/Annotation** which is not mandatory (in Python) and the code will execute either way. (that's a complementary convention)
**3) Code Formatting**, that is also conventional and don't affect the code execution (just code quality)
Beside, we have features like IntelliSense (code completion, suggestions, etc.) that is just a feature helps  speed-up code writing, right?

My intention is to: (your answers means the opposite intent)
use **ruff** to handle the no. 1, **Pyright** to covers no. 2 and **Black** to take care of no. 3 **(3 concerns, 3 tools, 3 settings)** and keep **Pylance as language server**, so install ruff and Black and set them as default tool to do their job, then I suppress Pylance features of no. 1 with `python.analysis.ignore` , the problem is Pyright Type-Checking disabled as a side effect (even when `typeCheckingMode` set to a value other than `off` ) this is a conflicting behavior of Pylance!
How come this is not the case when I setup **ruff + Black + mypy** (same functionalities, no hassle) ?
If you tight coupled Pyright to Pylance severely, means you forced it to either use them as a whole or just remove Pylance and use another Language Server (like the one astral/ruff provide)

**Reply by rchiodo** [reply]
Pylance doesn't have a way to turn off number 1 except through ignore. Ignore means ignore everything though. 

What you're asking for is only ignore syntax errors and those type checking rules that would actually cause an error at runtime. An ignoreRuffDuplicates setting I guess.

mypy doesn't have this either, it just has fewer type checking rules.

Really this is asking for something that doesn't duplicate output from Ruff.

**Reply by farzbood** [reply]
We're somehow on a same page now.
I'm aware that Pyright has more comprehensive Type-Ckecking and is faster than mypy (that's my motivation to employ that), the problem is Pyright can't be installed side by side with Pylance, as a standalone component to be used independently.
If there was a flag in VSCode to consider it that way (differentiate between the integrated component in Pylance and it's standalone version) would be a global resolution to similar situations (some extensions integrating an OSS component/library in their closed-source product)

**Reply by rchiodo** [reply]
I think if we had a setting to ignore syntax errors, the rest of the conflicts with Rust could be handled with custom report settings.

For example, this error:

![image](https://github.com/user-attachments/assets/7345efc8-c4e0-4d34-823f-d2fc0ee34df8)

Can be turned off in Pylance/Pyright with :

```jsonc
{
  "python.analysis.disableTaggedHints": true
}
```

**Reply by farzbood** [reply]
Sounds right

**Reply by heejaechang** [reply]
I am not sure why reporting syntax errors is considered linting. usually linting means things like code styling errors.

**Reply by rchiodo** [reply]
It's not reporting them that's the problem. It's the fact that Ruff reports them too. There's unnecessary duplication.

**Reply by heejaechang** [reply]
I see, so ruff let users to control which errors to be reported including syntax errors, but we don't is the problem?

**Reply by heejaechang** [reply]
hmmm.. ruff doesn't allow disabling syntax error either?

> The [E999](https://docs.astral.sh/ruff/rules/syntax-error/) rule was used to report syntax errors, allowing users to silence these reports via Ruff's disablement mechanisms. This release deprecates the [E999](https://docs.astral.sh/ruff/rules/syntax-error/) rule, which will be removed in a future release. Ruff will now always report all syntax errors it encounters, meaning users can no longer ignore them. The motivation for this change is to ensure users are always aware of syntax errors, as these prevent proper linting. By making syntax error reporting mandatory, we eliminate the risk of users missing these critical issues.

**Reply by rchiodo** [reply]
> I see, so ruff let users to control which errors to be reported including syntax errors, but we don't is the problem?

No, the problem is the user has both Ruff and Pylance at the same time. They would like to use Ruff for syntax errors and all the other errors Ruff reports. For everything else they want to use Pylance. It only matter that we can't disable syntax errors. The user wants Ruff to show syntax errors.
